### PR TITLE
Use a pool of per-thread cudnn handles for each device, updated

### PR DESCRIPTION
**Pull request by mcarilli**
Rebased version of https://github.com/pytorch/pytorch/pull/14861, hopefully addressing @ezyang's comments.

### CODE REVIEW

File: aten/src/ATen/cudnn/Handle.cpp [hunk 1]
```diff
@@ -17,8 +15,25 @@ struct Handle {
...
+  // the destructors of temporaries to call cudnnDestroy on the handle.
+  // We can achieve safety (for the narrow case of stashing within STL containers)
+  // by defining copy and move constructors that transfer cudnnDestroy
+  // responsibility to the latest constructed object.  This is NOT a substitute for
+  // full-blown reference counting, but reference counting may be overkill here.
+  // Another alternative is to wrap the saved Handles in unique_ptrs, i.e.,
+  // unordered_map<int, vector<unique_ptr<Handle>>> created_handles;
+  void transfer(Handle& rhs) {
+    handle = rhs.handle;
+    rhs.handle = NULL;
```
Comment (ezyang) [inline aten/src/ATen/cudnn/Handle.cpp]: nit: use nullptr here

File: aten/src/ATen/cudnn/Handle.cpp [hunk 2]
```diff
@@ -17,8 +15,25 @@ struct Handle {
...
+  }
+  Handle(Handle& rhs) { transfer(rhs); }
```
Comment (ezyang) [inline aten/src/ATen/cudnn/Handle.cpp]: This is not a thing. You can have copy constructor (`const Handle&`) or move constructor (`Handle&&`); a `Handle&` constructor is not a thing.

File: aten/src/ATen/cudnn/Handle.cpp [hunk 3]
```diff
@@ -17,8 +15,25 @@ struct Handle {
...
   }
+  // std::vector.emplace() and push_back() may route through temporaries and call
+  // copy/move constructors along the way.  If this is the case, we don't want
+  // the destructors of temporaries to call cudnnDestroy on the handle.
+  // We can achieve safety (for the narrow case of stashing within STL containers)
+  // by defining copy and move constructors that transfer cudnnDestroy
+  // responsibility to the latest constructed object.  This is NOT a substitute for
+  // full-blown reference counting, but reference counting may be overkill here.
+  // Another alternative is to wrap the saved Handles in unique_ptrs, i.e.,
+  // unordered_map<int, vector<unique_ptr<Handle>>> created_handles;
```
Comment (ezyang) [inline aten/src/ATen/cudnn/Handle.cpp]: I think, in less words, what you are trying to say is that `Handle` should be moveable, but not copyable. As I mentioned on chat, vector can handle move-only types perfectly well; just don't try to resize() a vector of those. Example: https://godbolt.org/z/WLbYje

File: aten/src/ATen/cudnn/Handle.cpp [hunk 4]
```diff
@@ -17,8 +15,25 @@ struct Handle {
...
+  void transfer(Handle& rhs) {
+    handle = rhs.handle;
+    rhs.handle = NULL;
+  }
+  Handle(Handle& rhs) { transfer(rhs); }
+  Handle(Handle&& rhs) { transfer(rhs); }
+  // operator= takes argument by value
+  Handle& operator=(Handle rhs) { transfer(rhs); return *this; }
```
Comment (ezyang) [inline aten/src/ATen/cudnn/Handle.cpp]: Careful. You have to check that rhs is not this; otherwise you'll leak the handle.

Comment (mcarilli) [reply aten/src/ATen/cudnn/Handle.cpp]: rhs is taken by value, following one of my (favorite stackoverflow posts)[https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom].
So within the function body, `rhs` will never be `*this`.  You could write something like
```
Handle handle_lvalue;
handle_lvalue = std::move(handle_lvalue);
```
but I think in this case `handle_lvalue` will still end up owning the original handle it constructed (the `rhs` argument to `operator=` will be move-constructed from `handle_lvalue`, and steal the handle, then within `operator=`, `handle_lvalue` will steal it back.
https://godbolt.org/z/cN2NQL
^ I don't think godbolt will let me run interactively, but when I compile and run on my local machine, it prints 1.

Comment (ezyang) [reply aten/src/ATen/cudnn/Handle.cpp]: Oh, I didn't realize you were using the copy-and-swap idiom here. I think you're the first user of this in PyTorch :) Worth a brief comment, I suppose.

File: aten/src/ATen/cudnn/Handle.cpp [hunk 5]
```diff
@@ -34,8 +49,79 @@ struct Handle {
...
+
+    if(available_handles[device].size() > 0)
+    {
+      my_handles[device] = available_handles[device].back();
+      available_handles[device].pop_back();
+    }
+    else
+    {
+      // In local testing, I do observe that emplace_back sometimes routes through temporaries
+      // that incur copy-constructor and destructor calls.  See comments in Handle above.
```
Comment (ezyang) [inline aten/src/ATen/cudnn/Handle.cpp]: Out of date now?

**Review by ezyang — APPROVED** [review]
OK. Let me know if you want to land this now, or wait for some straggler extra changes.

File: aten/src/ATen/cudnn/Handle.cpp [hunk 6]
```diff
@@ -12,8 +13,10 @@ namespace {
 
 struct Handle {
   cudnnHandle_t handle;
-  Handle() : handle(nullptr) {
-    AT_CUDNN_CHECK(cudnnCreate(&handle));
+  Handle(bool create = false) : handle(nullptr)
```
Comment (ezyang) [inline aten/src/ATen/cudnn/Handle.cpp]: nit: You could make this more explicit by using the named constructor idiom.

```
struct Handle {
  enum Create { CREATE; };
  Handle(Create) { ... }
  Handle() { ... }
}

Handle h(Handle::CREATE)
```

Comment (mcarilli) [reply aten/src/ATen/cudnn/Handle.cpp]: Neat, I didn't know about this one.

**Comment by ezyang** [comment]
@mcarilli Unfortunately, I think this directly caused a greatly increased incident of timeouts on

```
17:23:54 test_rnn_backward_to_input_but_not_parameters_cuda (__main__.TestAutograd) ... Build timed out (after 90 minutes). Marking the build as failed.
```

Sample job: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test2/18205//console

We saw a similar failure on the PR https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test2/18032/console but I saw that we had passed previously so I thought it was just flakiness. I'm unfortunately going to have to unland this.

**Pull request closed without merging**